src/Collapse/include/gudhi/Flag_complex_sparse_matrix.h

```diff
@@ -0,0 +1,376 @@
+  void insert_vertex(Vertex_handle vertex) {
+    auto n = row_to_vertex_.size();
+    auto result = vertex_to_row_.emplace(vertex, n);
+    // If it was not already inserted - Value won't be updated by emplace if it is already present
+    if (result.second) {
+      // Initializing the diagonal element of the adjency matrix corresponding to rw_b.
+      sparse_row_adjacency_matrix_[n].insert(n) = -1; // not an edge
+      // Must be done after reading its size()
+      row_to_vertex_.push_back(vertex);
+    }
+  }
+
+  // Insert an edge in the data structure
+  // @exception std::invalid_argument In debug mode, if u == v
+  void insert_new_edge(Vertex_handle u, Vertex_handle v, Row_index idx)
+  {
+    // The edge must not be added before, it should be a new edge.
+    insert_vertex(u);
+    GUDHI_CHECK((u != v),
+            std::invalid_argument("Flag_complex_sparse_matrix::insert_new_edge with u == v"));
+
+    insert_vertex(v);
+#ifdef DEBUG_TRACES
+    std::cout << "Insertion of the edge begins " << u <<", " << v << std::endl;
+#endif  // DEBUG_TRACES
+
+    auto rw_u = vertex_to_row_.find(u);
+    auto rw_v = vertex_to_row_.find(v);
+#ifdef DEBUG_TRACES
+    std::cout << "Inserting the edge " << u <<", " << v << std::endl;
+#endif  // DEBUG_TRACES
+    sparse_row_adjacency_matrix_[rw_u->second].insert(rw_v->second) = idx;
+    sparse_row_adjacency_matrix_[rw_v->second].insert(rw_u->second) = idx;
+  }
+
+ public:
+  /** \brief Flag_complex_sparse_matrix constructor from a range of filtered edges.
+   *
+   * @param[in] filtered_edge_range Range of filtered edges. Filtered edges must be in
+   * `Flag_complex_sparse_matrix::Filtered_edge`.
+   *
+   * There is no need the range to be sorted, as it will be performed in
+   * `Flag_complex_sparse_matrix::filtered_edge_collapse`.
+   */
+  template<typename Filtered_edge_range>
+  Flag_complex_sparse_matrix(const Filtered_edge_range& filtered_edge_range)
+  : f_edge_vector_(filtered_edge_range.begin(), filtered_edge_range.end()) {
+    // To get the number of vertices
+    std::unordered_set<Vertex_handle> vertices;
+    for (Filtered_edge filtered_edge : filtered_edge_range) {
+      Vertex_handle u;
+      Vertex_handle v;
+      std::tie(u,v) = std::get<0>(filtered_edge);
+      vertices.emplace(u);
+      vertices.emplace(v);
+    }
+    // Initializing sparse_row_adjacency_matrix_, This is a row-major sparse matrix.
+    sparse_row_adjacency_matrix_.resize(vertices.size(), Sparse_vector(vertices.size()));
+  }
+
+  /** \brief Flag_complex_sparse_matrix constructor from a proximity graph, cf. `Gudhi::compute_proximity_graph`.
+   *
+   * @param[in] one_skeleton_graph The one skeleton graph. The graph must be in
+   * `Flag_complex_sparse_matrix::Proximity_graph`.
+   *
+   * The constructor is computing and filling a vector of `Flag_complex_sparse_matrix::Filtered_edge`
+   */
+  Flag_complex_sparse_matrix(const Proximity_graph& one_skeleton_graph)
```

Comment:
Why hardcode the type of the graph instead of doing it like `Simplex_tree::insert_graph`?

Reply:
Done on b525bb68
Tell me if the documentation is ok.
I had to modify the constructor from filtered edges as it was conflicting.

Reply:
`end` doesn't point to the last element, it points one step after it.
That's one reason why I don't like designs with constructors so much, you cannot easily give 2 different names to 2 different constructors like you can with normal functions (member or not).
One possibility is to dispatch automatically depending on if the template type accepts begin/end, if boost::graph_traits is specialized, etc.
Another possibility is to add dummy parameters, so one would have for instance
```c++
typedef Flag_complex_edge_collapser<...> EC;
EC collapser(EC::from_range, the_range);
EC collapser2(EC::from_graph, the_graph);
```
Or we can make the constructor do almost nothing and have users call a function init_from_graph or init_from_edge_range.
Or as you did we can pass the range old-style.
Etc.
Let's forget about it for now, we'll come back to it later.

Reply:
And what about removing the constructor with `Filtered_edges` ?
The simplex_tree insert_graph has only one signature and it is ok.

Reply:
Ok, let's precise on 716085ac

Reply:
If we keep only one of them, I think the one taking a range is more important. With an EdgeListGraph, I can write a one-liner to give it the interface of a range:
```c++
boost::adaptors::transform(edges(graph), [&](auto&&x){return std::make_tuple(source(x, graph), target(x, graph), get(Gudhi::edge_filtration_t(), graph, x));})
```
while if I have a range and need to give it the interface of an EdgeListGraph, I need
- boost::adaptors::transform so that the value_type is a type with .first and .second
- boost::edge_list to wrap it with a graph interface
- associate a property to this graph, which seems to require different code depending if the range was random-access, using undocumented implementation details of edge_list.